File: Elixir/hoshimeguri2async.ex
# $ elixirc hoshimeguri2async.ex
# $ elixir -e 'Hoshimegrui2async.main' ../data.txt

defmodule Hoshimegrui2async do

  defmodule Solver do
    @star ~c(AHCJEBGDIF)

    def step(p, ?R) when rem(p, 2) == 1 do
      rem(p + 2, 10)
    end

    def step(p, ?W) when rem(p, 2) == 0 do
      rem(p + 8, 10)
    end

    def step(p, _) do
      rem(p + 1, 10)
    end

    def solve(input) do
      input_chars = String.to_char_list(input)
      first_index = Enum.find_index(@star, fn(c) -> c == hd(input_chars) end)
      tl(input_chars)
        |> Enum.reduce([first_index], fn(c, a) -> [step(hd(a), c)|a] end)
        |> Enum.reverse
        |> Enum.map(fn(i) -> Enum.at(@star, i) end)
        |> List.to_string
    end

    def solve do
      receive do
        {from, input, expected} -> send from, {input, expected, solve(input)}
      end
    end
  end

  defmodule Tester do
    def test(line) do
      [input, expected] = String.split(line, ~r{[ \n]}, trim: true)
      pid = spawn(&Solver.solve/0)
      send pid, {self, input, expected}
    end

    def judge(input, expected, actual) do
      if actual == expected do
        IO.write "."
      else
        IO.write """

        input:    #{input}
        expected: #{expected}
        actual:   #{actual}
        """
      end
    end

    def judge(0) do
      :ok
    end

    def judge(n) do
      receive do
        {input, expected, actual} ->
          judge(input, expected, actual)
          judge(n - 1)
      end
    end
  end

  defp read_lines(device) do
    line = IO.read(device, :line)
    case line do
      :eof -> []
      _    -> [line|read_lines(device)]
    end
  end

  def main do
    filename = hd(System.argv)
    {:ok, file} = File.open(filename)
    lines = read_lines(file)
    File.close file
    Enum.each lines, &Tester.test(&1)
    Tester.judge(length(lines))
    IO.puts("")
  end
end

File: Erlang/hoshimeguri2async.erl
% ビルド:
%   $ erlc hoshimeguri2async.erl
%
% 実行: 二つのコンソールを開いて、まず solver ノードを起動し、次に tester ノードを実行する
%   $ erl -sname solver
%   $ erl -run hoshimeguri2async main --noshell -sname tester -node solver@<hostname> -data ../data.txt

-module(hoshimeguri2async).
-export([main/0, solver/0]).

index(C) -> string:str("AHCJEBGDIF", [C]) - 1.

step(P0, $R) when P0 rem 2 =:= 1 -> (P0 + 2) rem 10;
step(P0, $W) when P0 rem 2 =:= 0 -> (P0 + 8) rem 10;
step(P0, _ ) -> (P0 + 1) rem 10.

solve([FirstPoint|Colors]) ->
  Indices = lists:foldl(
    fun(Color, Indices) ->
      lists:append(Indices, [step(lists:last(Indices), Color)])
    end,
    [index(FirstPoint)],
    Colors
  ),
  lists:map(fun(I) -> lists:nth(I + 1, "AHCJEBGDIF") end, Indices).

loop() ->
  receive
    {From, Input, Expected} ->
      Actual = solve(Input),
      From ! {Input, Expected, Actual},
      loop()
  end.

solver() ->
  case whereis(solver) of
    undefined -> Pid = spawn(node(), fun loop/0), register(solver, Pid), Pid;
    Pid       -> Pid
  end.

judge(_, Expected, Actual) when Expected == Actual ->
  io:put_chars(".");
judge(Input, Expected, Actual) ->
  io:format("""

input:    ~s
expected: ~s
actual:   ~s
""",
    [Input, Expected, Actual]
  ).

judge(0) -> ok;
judge(N) ->
  receive
    {Input, Expected, Actual} ->
      judge(Input, Expected, Actual),
      judge(N - 1)
  end.

test(Solver, Lines) ->
  lists:map(
    fun(Line) ->
      [Input, Expected] = string:tokens(Line, " \n"),
      Solver ! {self(), Input, Expected}
    end,
    Lines
  ),
  judge(length(Lines)).

read_lines(Handle) ->
  case file:read_line(Handle) of
    {ok, Line} -> [Line|read_lines(Handle)];
    _Otherwise -> []
  end.

main() ->
  {ok, [[Node]]} = init:get_argument(node),
  {ok, [[Filename]]} = init:get_argument(data),
  {ok, Handle} = file:open(Filename, read),
  Lines = read_lines(Handle),
  file:close(Handle),
  case rpc:call(list_to_atom(Node), module_info(module), solver, [], 1000) of
    {badrpc, Reason} -> io:format("~p~n", [Reason]);
    Solver -> test(Solver, Lines)
  end,
  io:nl(),
  halt().
